### PR TITLE
Update _widget_reader.py

### PR DESCRIPTION
**Pull request by TomasPetro**
I've edited it and it should work (I've tested saving the data and then loading it from File->OpenFile).  I did this for YtReader and TimeSeriesReader. But the validate_data_model function does not return the model only the JSON. I don't know if it bothers or not but I had a problem converting the returned model to JSON. Besides, there is a hardcoded $schema$ because without it I couldn't load JSON files and I didn't know where I could get it into the function.

### CODE REVIEW

**Comment by chrishavlin** [comment]
Thanks, @TomasPetro ! I should have time to look this over later in the week.

File: src/yt_napari/_widget_reader.py [hunk 1]
```diff
@@ -94,6 +97,7 @@ def remove_selection(self):
 
 class ReaderWidget(YTReader):
     _pydantic_model = _data_model.DataContainer
+    app = QApplication(sys.argv)
```
Comment (chrishavlin) [inline src/yt_napari/_widget_reader.py:100]: These lines shouldn't be added -- napari takes care of all QT application and context handling and are why tests are stalling.
```suggestion
```

File: src/yt_napari/_widget_reader.py [hunk 2]
```diff
@@ -106,6 +110,27 @@ def add_load_group_widgets(self):
...
+        file_dialog.setAcceptMode(QFileDialog.AcceptSave)
+        file_dialog.setNameFilter("JSON Files (*.json);;All Files (*)")
+
+        if file_dialog.exec_():
+            file_path = file_dialog.selectedFiles()[0]
+            if file_path:
+                # Save the JSON data to the selected file
+                with open(file_path, 'w') as json_file:
+                    json.dump(py_kwargs, json_file, indent=4)    
+    app.exec_()
```
Comment (chrishavlin) [inline src/yt_napari/_widget_reader.py:132]:
```suggestion
```

File: src/yt_napari/_widget_reader.py [hunk 3]
```diff
@@ -213,6 +243,7 @@ def get_current_pydantic_kwargs(self) -> dict:
 
 class TimeSeriesReader(YTReader):
     _pydantic_model = _data_model.Timeseries
+    app = QApplication(sys.argv)
```
Comment (chrishavlin) [inline src/yt_napari/_widget_reader.py:246]:
```suggestion
```

File: src/yt_napari/_widget_reader.py [hunk 4]
```diff
@@ -223,7 +254,51 @@ def add_load_group_widgets(self):
...
+        file_dialog.setAcceptMode(QFileDialog.AcceptSave)
+        file_dialog.setNameFilter("JSON Files (*.json);;All Files (*)")
+
+        if file_dialog.exec_():
+            file_path = file_dialog.selectedFiles()[0]
+            if file_path:
+                # Save the JSON data to the selected file
+                with open(file_path, 'w') as json_file:
+                    json.dump(py_kwargs, json_file, indent=4)    
+    app.exec_()
```
Comment (chrishavlin) [inline src/yt_napari/_widget_reader.py:277]:
```suggestion
```

**Review by chrishavlin — CHANGES_REQUESTED** [review]
I was trying to figure out why the tests are stalling -- it's cause you added the QApplication instances. Napari handles all the main QT context, these need to be removed.

**Comment by chrishavlin** [comment]
also, if you type `pre-commit.ci autofix` on it's own in a comment then the pre-commit bot will try to fix the failing style checks.

**Comment by TomasPetro** [comment]
I looked at those comments and yes they were unnecessary features. I don't know why I put them there. I deleted them and tried them on my pc and everything went as before.

File: src/yt_napari/_widget_reader.py [hunk 5]
```diff
@@ -254,30 +323,12 @@ def load_data(self):
 
         # now ready to instantiate the base model
         py_kwargs = {
+            "$schema": "https://yt-napari.readthedocs.io/en/latest/_static/yt-napari_latest.json",
```
Comment (chrishavlin) [inline src/yt_napari/_widget_reader.py:326]: Same change here:
```suggestion
            "$schema": schema_name,
```

**Review by chrishavlin — CHANGES_REQUESTED** [review]
Thanks for the nice work, @TomasPetro ! 

I left some comments with some small changes to how you're setting the `"$schema"` entry. 

You'll also need to fix the failing style checks and add a new test.

**style checks**: you can read the full log here https://results.pre-commit.ci/run/github/448138442/1699393334.AOajljEzRPyJ5O8Rgft3mQ and manually fix the failures, or you can also type `pre-commit.ci autofix` on its own in a comment and the bot will try to fix it for you (you'd then want to pull the changes locally). 

**new test**: I did checkout your branch locally to make sure it works (and it does!) but when adding new features we also need to add to the automated test suite. So this PR will need a new test in `src/yt_napari/_tests/test_widget_reader.py`. Adding tests for widget features can be a little tricky -- but you should be able to follow how the other tests are written to write a new one. In this case, I think the easiest route would be to monkeypatch the `QFileDialog` `exec_()` and `selectedFiles()` functions (see [this link](https://docs.pytest.org/en/7.1.x/how-to/monkeypatch.html) for how to monkeypatch with pytest).

**Comment by TomasPetro** [comment]
@chrishavlin 
So I installed tox but I'm currently having trouble running it. Tried to run it in different folders but it keeps throwing an error. Mostly similar. Are there other ways to run the tests or am I getting an error somewhere when running it? 
image

**Comment by chrishavlin** [comment]
Oh, sorry, I must have some docs out of date. I don't really use tox anymore with `yt-napari` -- I just run `pytest -v .` to run the tests in my current environment and then rely on the github CI to make sure it passes with other systems and python versions.

**Comment by TomasPetro** [comment]
Previously I tried pytest but i got there problem with qtbot. And even in those tests I didn't touch. In total I get 12 such errors. Do you know of any solution to this problem ? 
image

**Comment by chrishavlin** [comment]
you'll need the `pytest-qt` plugin as well, `pip install pytest-qt`

**Comment by TomasPetro** [comment]
Hi. 
I'm currently working on tests but I've been troubleshooting the QfileDialog window that pops up when I save a json file for a few days.
 I've tried qtbot etc but it didn't help and I still need to manually type the filename and then save. Do you know of any suitable implementation I could try ? 
image

**Comment by chrishavlin** [comment]
I think the easiest approach will be to monkeypatch within pytest it so that you don't have to deal with the popup, see my comment up higher:

> In this case, I think the easiest route would be to monkeypatch the QFileDialog exec_() and selectedFiles() functions (see [this link](https://docs.pytest.org/en/7.1.x/how-to/monkeypatch.html) for how to monkeypatch with pytest).

So you should be able to write a test that provides string inputs to provide a temporary filename and avoid the pop-up entirely.

**Comment by TomasPetro** [comment]
pre-commit.ci autofix

**Comment by TomasPetro** [comment]
Hi, so I wrote the tests. Will you take a look at it ?  Currently it doesn't pass only codecs/patches but I didn't do anything with those functions on those lines.

**Comment by chrishavlin** [comment]
Great! Thanks @TomasPetro ! I should be able to look later today or tomorrow.

**Review by chrishavlin — APPROVED** [review]
Looks great, thanks @TomasPetro ! I'll push up some changes to fix the coverage issues (as well as another couple small changes to the tests) and then this should be ready to merge.

**Comment by TomasPetro** [comment]
Nice :)

**Pull request merged**